src/charm.py

```diff
@@ -86,14 +87,18 @@ def __init__(self, *args):
         # TODO:
         #  ingress route provisioning a separate TCP ingress for each receiver if GRPC doesn't work directly
 
-        self._ingress = IngressPerAppRequirer(
-            self, port=self.tempo.tempo_server_port, strip_prefix=True
-        )
-
+        self._ingress = TraefikRouteRequirer(self, self.model.get_relation("ingress"), "ingress")  # type: ignore
         self._tracing = TracingEndpointProvider(
-            self, host=self.tempo.host, external_url=self._ingress.url
+            self, host=self.tempo.host, external_url=self._ingress.external_host
         )
 
+        self.framework.observe(self.on["ingress"].relation_joined, self._configure_ingress)
+        self.framework.observe(self.on.leader_elected, self._configure_ingress)
```

Comment:
all observe calls should be: 

`self.on.event, self._on_event`

and in `def self._on_event` you can do whatever you like. This is easier to maintain and follow.

Reply:
secondly, why do we need to observe the leader-elected hook? 
I guess it's because traefik-route goes through app data, and if your leader dies and you are elected, you need to take lead of traefik-route. Can you add a comment?

Reply:
Changed it according to the guidelines and added the requested comment.